text/0000-debugger-visualizer.md

```diff
@@ -0,0 +1,643 @@
+    > locs: {...}
+      > __0: (4) vec![None, None, None, None]
+      > named_groups: (refs:2) size=0, capacity=1
+        > [raw]: alloc::sync::Arc<std::collections::hash::map::HashMap<...>>
+          > ptr: {pointer:0xNNNNNNNN}
+            > pointer: {...}
+              > strong: {...}
+              > weak: {...}
+              > data: size=0, capacity=1
+                > base: {...}
+                  > hash_builder: {...}
+        ...
+```
+
+The debugger shows the structure of the data, not its meaning. It is not very
+useful for Carol. Even the implementor of `regex` would have a hard time knowing
+how to decode this. In reality, when trying to understand the state of the captures
+variable there are several methods defined for a `Captures` type that paint
+the actual picture in terms of the information a Rust developer would like to
+extract from this variable. In order to meaningfully understand what the `Captures`
+type is truly trying to tell us, it would be very helpful to visualize this data
+differently in the debugger.
+
+What we _want_ is something like this:
+
+```text
+> Variables:
+  > captures: {...}
+    > 1: "SOME_CONSTANT"
+    > 2: "42"
+    > 3: "// some developer comment"
+```
+
+This RFC will describe how to support adding Natvis as well as GDB's pretty printers.
+
+Natvis is supported by:
+
+* The Windows Debugger (WinDbg)
+* Visual Studio Debugger
+
+Pretty printers are supported by:
+
+* GDB
+* LLDB
+
+It should be easy for Rust developers to add debugger visualizations to their
+crates.
+
+## Supporting Natvis
+
+This section describes how Microsoft's Natvis is supported in Rust.
+
+To use Natvis, developers write XML documents that describe how debugger types
+should be displayed using the natvis schema. (See: https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects?view=vs-2019)
+The Natvis files provide patterns, which match type names, and for matching
+types, a description of how to display those types. This allows for some
+limited support for generic types.
+
+Rust developers can add one or more `.natvis` files to their crate. Through
+the use of a new Rust attribute, `#[debugger_visualizer]`, the compiler will
+encode the contents of the `.natvis` file in the crate metadata if the target
+is an `rlib`. If the target is a `dll` or `exe`, the `/NATVIS` MSVC linker flag is
+set for each `.natvis` file which will embed the Natvis visualizations into the PDB.
+
+To provide Natvis files, developers create a file with the `.natvis` file
+extension.
+
+See the below example for how rustc would embed these Natvis files in the
+debuginfo for a binary.
+
+As an example, consider a crate `foo` with this directory structure:
+
+```text
+/Cargo.toml
+/Foo.natvis (Note: the .natvis file does not have to match the name of the crate.)
+  +-- src
+      +-- main.rs
+```
+
+Where `main.rs` contains:
+
+```rust
+#[debugger_visualizer(natvis_file = "../Foo.natvis")]
```

Comment:
This would be 
```
#![debugger_visualizer(natvis_file = "../Foo.natvis")]
 ^
// Note the `!`
```
right? That is, it would be a top-level attribute on the crate rather than on types?

Reply:
In some instances yes. The `#[debugger_visualizer]` attribute is not a `CrateLevel` attribute. Currently the only restriction I have put on this is that it targets modules. This way it may be used in any Rust source file using the `#![debugger_visualizer(natvis_file = "../Foo.natvis")]` syntax at the top of a file or 

```
#[debugger_visualizer(natvis_file = "../Foo.natvis")]
mod foo;
```

using the attribute directly on a module.

Thus this attribute isn't solely on the top-level crate. The RFC does not clarify this so I will update the RFC to clarify that it targets modules and can be used in either fashion.